pandas/tools/merge.py

```diff
@@ -112,44 +183,160 @@ def ordered_merge(left, right, on=None, left_by=None, right_by=None,
         of DataFrame.
     """
     def _merger(x, y):
+        # perform the ordered merge operation
         op = _OrderedMerge(x, y, on=on, left_on=left_on, right_on=right_on,
-                           # left_index=left_index, right_index=right_index,
-                           suffixes=suffixes, fill_method=fill_method)
+                           suffixes=suffixes, fill_method=fill_method,
+                           how=how)
         return op.get_result()
 
     if left_by is not None and right_by is not None:
         raise ValueError('Can only group either left or right frames')
     elif left_by is not None:
-        if not isinstance(left_by, (list, tuple)):
-            left_by = [left_by]
-        pieces = []
-        for key, xpiece in left.groupby(left_by):
-            merged = _merger(xpiece, right)
-            for k in left_by:
-                # May have passed ndarray
-                try:
-                    if k in merged:
-                        merged[k] = key
-                except:
-                    pass
-            pieces.append(merged)
-        return concat(pieces, ignore_index=True)
+        result, _ = _groupby_and_merge(left_by, left, right,
+                                       lambda x, y: _merger(x, y))
     elif right_by is not None:
-        if not isinstance(right_by, (list, tuple)):
-            right_by = [right_by]
-        pieces = []
-        for key, ypiece in right.groupby(right_by):
-            merged = _merger(left, ypiece)
-            for k in right_by:
-                try:
-                    if k in merged:
-                        merged[k] = key
-                except:
-                    pass
-            pieces.append(merged)
-        return concat(pieces, ignore_index=True)
+        result, _ = _groupby_and_merge(right_by, right, left,
+                                       lambda x, y: _merger(y, x))
     else:
-        return _merger(left, right)
+        result = _merger(left, right)
+    return result
+
+
+def asof_merge(left, right, on=None,
+               left_on=None, right_on=None,
+               by=None,
+               suffixes=('_x', '_y'),
+               tolerance=None,
+               allow_exact_matches=True):
+    """Perform an asof merge. This is similar to a left-join except that we
+    match on nearest key rather than equal keys.
+
+    For each row in the left DataFrame, we select the last row in the right
+    DataFrame whose 'on' key is less than the left's key. Both DataFrames must
+    be sorted by the key.
+
+    Optionally perform group-wise merge. This matches the 'by' key equally,
+    in addition to the nearest match on the 'on' key.
+
+    .. versionadded 0.18.2
+
+    Parameters
+    ----------
+    left: DataFrame
+    right: DataFrame
+        Data will be de-duplicated on the [by, on] combination if needed.
+    on: label or list
+        Field names to join on. Must be found in both DataFrames.
```

Comment:
I don't think it is true. It just works without specifying `on` (if there is a common column name). So we should either document this or disallow (and test in both cases). For consistency with merge, I would just allow it (as it is at the moment).